clang/lib/Basic/FileManager.cpp

```diff
@@ -472,9 +472,13 @@ FileEntryRef FileManager::getVirtualFileRef(StringRef Filename, off_t Size,
 }
 
 OptionalFileEntryRef FileManager::getBypassFile(FileEntryRef VF) {
+  // Compute the requested name and make it an absolute path.
+  SmallString<256> Path(VF.getNameAsRequested());
+  makeAbsolutePath(Path);
```

Comment:
Why do we need `makeAbsolutePath` here?

Reply:
There is a test to check a.pcm and /path/a.pcm both works, maybe `clang/test/Modules/explicit-build-relpath.cpp`
Ideally, this should be realpath, but that is going to be expensive. This is just to make the bad workaround less bad.

Reply:
@jansvoboda11 any ideas here?

Reply:
That's curious. @cachemeifyoucan can you explain a bit more why this is necessary?

`FileManager::getFileRef()` does not make the requested path absolute and stores it as-is in the returned `FileEntryRef`. This means that we've been able to successfully use the relative path to obtain `VF`. Why exactly can't we use the same relative path (`VF.getNameAsRequested()`) here in `getBypassFile()`?

Reply:
Didn't try and didn't investigate deep into the failure, just workaround the behavior difference.

My guess is B.pcm is built with relative path A.pcm, and A.pcm is also imported via absolute path. They need to be the same FileEntryRef to avoid duplicated module error.

Reply:
It will be hard to come up with better solution without knowing what exactly goes wrong in the test case. But In general, the intent of `FileManager::getBypassFile()` is to provide new `FileEntry` that's **not** de-duplicated based on the inode. If that's causing duplicate module errors down the line, it seems we'll need to fix the client.

If we make the path absolute here, we'll be returning a `FileEntryRef` that could potentially report different name compared to the original, which is surprising.

Reply:
Current `getBypassFile` implementation is totally diverted from what the comments said. I really want to revert the workaround since the workaround is absolutely wrong, this change is just some middle ground that makes module work in more cases.

Reply:
How does it diverge? Do you mean the fact that it does de-duplicate bypass files based on `VF.getName()`?

What's your plan for reverting the workaround for ext4 recycling inodes?

Reply:
getBypassFile is not supposed to do any dedup according to comments but it is already deduping. After the change, getBypassFile will only dedup from absolute path (not realpath), which is not unreasonable, or more unreasonable than the semantic before the patch.

I don't have plan to revert the change (just want to). All I can do is putting the workaround into a less horrible place.

Reply:
Can we have the returned FileEntryRef use the requested path and only use the absolute path for dedup? That would be closer to how normal FileEntryRef lookup works.  I'm not so concerned about the dedup, I'm only concerned about us changing the path in the FileEntryRef.

Reply:
The requested path is just key for dedup. I am not sure if I want to add another map for dedup only. I think we either keep makeAbsolute, or XFAIL that test.

Reply:
I decide to put the makeAbsolutePath behind the same guard so it is not changing behavior when module bypass is not used.